pyhap/loader.py

```diff
@@ -8,95 +8,86 @@
                 'UUID' not in char_dict:
             raise KeyError('Could not load char {}!'.format(name))
         return Characteristic.from_dict(name, char_dict)
 
-    def get_service(self, name, char_loader=None):
-        """Return new service object.
-
-        :param char_loader: `TypeLoader` object to use when creating the
-            characteristics for adding to instantiated service.
-        :type char_loader: TypeLoader
-
-        :raise KeyError: When service file did not contain necessary keys.
-        """
-        char_loader = char_loader or get_char_loader()
-        service_dict = self.get(name)
+    def get_service(self, name):
+        """Return new service object."""
+        service_dict = self.serv_types[name].copy()
         if 'RequiredCharacteristics' not in service_dict or \
                 'UUID' not in service_dict:
             raise KeyError('Could not load service {}!'.format(name))
-        return Service.from_dict(name, service_dict, char_loader)
+        return Service.from_dict(name, service_dict, self)
 
+    @classmethod
+    def from_dict(cls, char_dict=None, serv_dict=None):
+        """Create a new instance directly from json dicts."""
+        loader = cls.__new__(Loader)
+        loader.char_types = char_dict or {}
+        loader.serv_types = serv_dict or {}
+        return loader
 
-def get_char_loader(desc_file=CHARACTERISTICS_FILE):
-    """Get a CharacteristicLoader with characteristic descriptions in the given file.
 
-    Uses a 'singleton' when the file is `CHARACTERISTICS_FILE`.
+def get_loader():
+    """Get a service and char loader.
+
+    If already initialized it returns the existing one.
     """
-    global _char_loader
-    if desc_file == CHARACTERISTICS_FILE:
-        if _char_loader is None:
-            with open(desc_file, 'r') as fp:
-                _char_loader = TypeLoader(fp)
-        return _char_loader
+    global _loader
+    if _loader is None:
+        _loader = Loader()
+    return _loader
+
 
-    with open(desc_file, 'r') as fp:
-        ld = TypeLoader(fp)
-    return ld
+def get_char_loader(desc_file=None):
+    """Get a CharacteristicLoader with characteristic descriptions in the given file.
+
+    .. deprecated:: 2.0
+       Use `get_loader` instead.
+    """
+    logger.warning(
+        "'get_char_loader' is deprecated. Use 'get_loader' instead.")
 
 
-def get_serv_loader(desc_file=SERVICES_FILE):
+def get_serv_loader(desc_file=None):
     """Get a ServiceLoader with service descriptions in the given file.
 
-    Uses a 'singleton' when the file is `SERVICES_FILE`.
+    .. deprecated:: 2.0
+       Use `get_loader` instead.
     """
-    global _serv_loader
-    if desc_file == SERVICES_FILE:
-        if _serv_loader is None:
-            with open(desc_file, 'r') as fp:
-                _serv_loader = TypeLoader(fp)
-        return _serv_loader
-
-    with open(desc_file, 'r') as fp:
-        ld = TypeLoader(fp)
-    return ld
+    logger.warning(
```

Comment:
This caller will then fail. Can we `return get_loader()` here and for the char?

Reply:
Yea, missed that